src/search/landmarks/landmark_status_manager.cc

```diff
@@ -1,171 +1,223 @@
+                /*
+                  We assume here that no natural orderings A->B are generated
+                  such that B holds in the initial state. Such orderings would
+                  only be valid in unsolvable problems.
+                */
+                assert(parent.second <= EdgeType::REASONABLE);
+                /*
+                  Reasonable orderings A->B for which both A and B hold in the
+                  initial state are immediately satisfied. We assume such
+                  orderings are not generated in the first place.
+                */
+                assert(!parent.first->get_landmark().is_true_in_state(
+                           initial_state));
+                utils::unused_variable(parent);
+                fut.set(id);
             }
         } else {
-            for (const FactPair &fact : landmark.facts) {
-                if (initial_state[fact.var].get_value() == fact.value) {
-                    reached.set(lm_node->get_id());
-                    ++inserted;
-                    break;
-                }
-            }
+            past.reset(id);
+            fut.set(id);
         }
     }
-    if (log.is_at_least_normal()) {
-        log << inserted << " initial landmarks, "
-            << num_goal_lms << " goal landmarks" << endl;
-    }
 }
 
-bool LandmarkStatusManager::process_state_transition(
+void LandmarkStatusManager::progress(
     const State &parent_ancestor_state, OperatorID,
     const State &ancestor_state) {
     if (ancestor_state == parent_ancestor_state) {
         // This can happen, e.g., in Satellite-01.
-        return false;
+        return;
     }
 
-    const BitsetView parent_reached = get_reached_landmarks(parent_ancestor_state);
-    BitsetView reached = get_reached_landmarks(ancestor_state);
-
-    int num_landmarks = lm_graph.get_num_landmarks();
-    assert(reached.size() == num_landmarks);
-    assert(parent_reached.size() == num_landmarks);
-
-    /*
-       Set all landmarks not reached by this parent as "not reached".
-       Over multiple paths, this has the effect of computing the intersection
-       of "reached" for the parents. It is important here that upon first visit,
-       all elements in "reached" are true because true is the neutral element
-       of intersection.
-
-       In the case where the landmark we are setting to false here is actually
-       achieved right now, it is set to "true" again below.
-    */
-    reached.intersect(parent_reached);
-
-
-    // Mark landmarks reached right now as "reached" (if they are "leaves").
-    for (int id = 0; id < num_landmarks; ++id) {
-        if (!reached.test(id)) {
-            LandmarkNode *node = lm_graph.get_node(id);
-            if (node->get_landmark().is_true_in_state(ancestor_state)) {
-                if (landmark_is_leaf(*node, reached)) {
-                    reached.set(id);
+    ConstBitsetView parent_past =
+        get_past_landmarks_const(parent_ancestor_state);
```

Comment:
The more idiomatic way to handle different const-ness and allow using a const version when we have non-const access is to have both methods called get_past_landmarks (etc.) and implement an automatic conversion from BitsetView to ConstBitsetView.

Reply:
By automatic conversion, do you mean adding a constructor `ConstBitsetView::ConstBitsetView(const BitsetView &other)`? I just tried that and it works, but it leads to some other ugliness. For starters, this constructor requires `ConstBitsetView` to have access to `BitsetView`'s private members. I solved this with a fried declaration (which required me to forward declare `ConstBitsetView`). But then we need to do the entire thing again for `(Const)ArrayView` since this is one of the members of `(Const)BitsetView.

I can push that change so you can see it better, but I first wanted to ask if this is roughly what you had in mind.

Reply:
There are two ways to implement automatic conversions from class A to class B: class B can have a non-explicit constructor from class A, or class A can implement an "operator B". In the latter case, you would have access to the private members of class A. So in this case, BitsetView would implement `ConstBitsetView operator ConstBitsetView() const {return ...;}`. It would probably be a good idea for the implementation to be in the header file so that it is inlinable and there is no performance cost compared to the current code.

Reply:
If you want to speed this up, we can look at the code together over Zoom (applies to both of you).

Reply:
@salome-eriksson Clemens and I looked at this together, and Clemens will push a commit.